File: Capítulo_8/Sondagem linear/Hash Table Linear Probing/Método put/Código em  Python/put.py
# defaultToString:
def defaul_to_string(item):
    if item is None:
        return "NONE"
    elif isinstance(item, str):
        return f"{item}"
    return str(item)

# Classe ValuePair
class ValuePair:
    def __init__(self, key, value):
        self.key = key
        self.value = value

    def __str__(self):
        return f"[#{self.key}: {self.value}]"
    
# Classe HashTableLinearProbing
class HashTableLinearProbing:
    def __init__(self, to_str_fn = defaul_to_string):
        self.to_str_fn = to_str_fn
        self.table = {}

    # Método loseloseHashCode
    def loselose_hash_code(self, key):
        if isinstance(key, (int, float)):
            return key
        table_key = self.to_str_fn(key)
        hash = 0
        for i in range(len(table_key)):
            hash += ord(table_key[i])
        return hash % 37
    
    # Método hash_code
    def has_code(self, key):
        return self.loselose_hash_code(key)
    
    # Método put
    def put(self, key, value):
        if key and value is not None:
            position = self.hash_code(key)
            table = self.get_table()
            if position not in table:
                self.set_table(position, ValuePair(key, value))
            else:
                index = position + 1
                while index in table:
                    index += 1
                self.set_table(index, ValuePair(key, value))
            return True
        return False

File: Capítulo_8/Sondagem linear/Hash Table Linear Probing/Método put/put.js
import { defaultToString, ValuePair } from "../../../util/util";

// Classe HashTableLinearProbing
class HashTableLinearProbing {
    constructor(toStrFn =  defaultToString) {
        this.toStrFn = toStrFn;
        this.table = {};
    }

    // Método loseloseHashCode
    loseloseHashCode(key) {
        if (typeof key === "number") {
            return key;
        }
        const tableKey = this.toStrFn(key);
        let hash = 0;
        for (let i = 0; i < tableKey.length; i++) {
            hash += tableKey.charCodeAt(i);
        }
        return hash % 37;
    }

    // Método hashCode
    hashCode(key) {
        return this.loseloseHashCode(key);
    }

    // Método put
    put(key, value) {
        if (key != null && value != null) {
            const position = this.hashCode(key);
            if (this.table[position == null]) {
                this.table[position] = new ValuePair(key, value);
            } else {
                let index = position + 1;
                while (this.table[index] != null) {
                    index++;
                }
                this.table[index] = new ValuePair(key, value);
            }
            return true;
        }
        return false;
    }
}